File: src/realtime-provider/migration-state.ts
/**
 * Durable migration checkpoint for realtime-provider rollout.
 * Survives restarts — lives under workspace/state/.
 */

import { existsSync, mkdirSync, readFileSync, renameSync, writeFileSync } from 'node:fs';
import { join } from 'node:path';
import { resolveWorkspace } from '../workspace_default.js';
import type {
	PhaseRecord,
	PhaseStatus,
	RealtimeProviderMigrationState,
	RealtimeProviderId,
} from './types.js';

export const MIGRATION_SCHEMA_VERSION = 1;
export const MIGRATION_FILENAME = 'realtime-provider-migration.json';

const PHASE_IDS = ['0', '1', '2', '3', '4'] as const;

export function migrationStatePath(workspace?: string): string {
	const ws = workspace || resolveWorkspace();
	return join(ws, 'state', MIGRATION_FILENAME);
}

function defaultState(): RealtimeProviderMigrationState {
	const phases: Record<string, PhaseRecord> = {};
	for (const id of PHASE_IDS) {
		phases[id] = { status: id === '0' ? 'complete' : 'pending' };
	}
	return {
		schema_version: MIGRATION_SCHEMA_VERSION,
		updated_at: new Date().toISOString(),
		current_phase: 1,
		phases,
		rollback: {
			provider: 'gemini',
			use_factory: true,
			vision_adapter: false,
		},
	};
}

export function readMigrationState(workspace?: string): RealtimeProviderMigrationState {
	const path = migrationStatePath(workspace);
	if (!existsSync(path)) {
		return defaultState();
	}
	try {
		const raw = JSON.parse(readFileSync(path, 'utf-8')) as RealtimeProviderMigrationState;
		if (raw.schema_version !== MIGRATION_SCHEMA_VERSION) {
			return { ...defaultState(), ...raw, schema_version: MIGRATION_SCHEMA_VERSION };
		}
		return raw;
	} catch {
		return defaultState();
	}
}

export function writeMigrationState(
	state: RealtimeProviderMigrationState,
	workspace?: string,
): void {
	const path = migrationStatePath(workspace);
	mkdirSync(join(path, '..'), { recursive: true });
	const next: RealtimeProviderMigrationState = {
		...state,
		updated_at: new Date().toISOString(),
	};
	const tmp = `${path}.tmp-${process.pid}`;
	writeFileSync(tmp, JSON.stringify(next, null, 2) + '\n');
	renameSync(tmp, path);
}

export function markPhase(
	phase: number | string,
	status: PhaseStatus,
	notes?: string,
	workspace?: string,
): RealtimeProviderMigrationState {
	const state = readMigrationState(workspace);
	const key = String(phase);
	const prev = state.phases[key] || { status: 'pending' as PhaseStatus };
	const now = new Date().toISOString();
	const record: PhaseRecord = {
		...prev,
		status,
		notes: notes ?? prev.notes,
	};
	if (status === 'in_progress' && !record.started_at) record.started_at = now;
	if (status === 'complete' || status === 'rolled_back') record.completed_at = now;
	state.phases[key] = record;
	if (status === 'complete') {
		const n = Number(phase);
		if (!Number.isNaN(n) && n >= state.current_phase) {
			state.current_phase = Math.min(n + 1, 4);
		}
	}
	writeMigrationState(state, workspace);
	return state;
}

export function recordRollbackSnapshot(
	provider: RealtimeProviderId,
	useFactory: boolean,
	visionAdapter: boolean,
	workspace?: string,
): void {
	const state = readMigrationState(workspace);
	state.rollback = { provider, use_factory: useFactory, vision_adapter: visionAdapter };
	writeMigrationState(state, workspace);
}

export function initMigrationStateIfMissing(workspace?: string): RealtimeProviderMigrationState {
	const path = migrationStatePath(workspace);
	if (existsSync(path)) return readMigrationState(workspace);
	const state = defaultState();
	state.phases['0'] = {
		status: 'complete',
		completed_at: new Date().toISOString(),
		notes: 'Vendor spikes: test-qwen-realtime-tools.py',
	};
	writeMigrationState(state, workspace);
	return state;
}

File: src/realtime_provider/migration_state.py
"""Durable migration checkpoint for realtime-provider rollout (Python mirror)."""

from __future__ import annotations

import json
import os
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

MIGRATION_SCHEMA_VERSION = 1
MIGRATION_FILENAME = "realtime-provider-migration.json"
PHASE_IDS = ("0", "1", "2", "3", "4")


def _repo_src() -> Path:
    return Path(__file__).resolve().parent.parent


def migration_state_path(state_dir: Path | None = None) -> Path:
    if state_dir is None:
        try:
            from workspace_default import resolve_workspace

            state_dir = resolve_workspace() / "state"
        except ImportError:
            state_dir = _repo_src().parent / "workspace" / "state"
    return state_dir / MIGRATION_FILENAME


def default_state() -> dict[str, Any]:
    phases = {pid: {"status": "complete" if pid == "0" else "pending"} for pid in PHASE_IDS}
    return {
        "schema_version": MIGRATION_SCHEMA_VERSION,
        "updated_at": datetime.now(timezone.utc).isoformat(),
        "current_phase": 1,
        "phases": phases,
        "rollback": {
            "provider": "gemini",
            "use_factory": True,
            "vision_adapter": False,
        },
    }


def read_migration_state(state_dir: Path | None = None) -> dict[str, Any]:
    path = migration_state_path(state_dir)
    if not path.is_file():
        return default_state()
    try:
        raw = json.loads(path.read_text())
        if raw.get("schema_version") != MIGRATION_SCHEMA_VERSION:
            merged = default_state()
            merged.update(raw)
            merged["schema_version"] = MIGRATION_SCHEMA_VERSION
            return merged
        return raw
    except (json.JSONDecodeError, OSError):
        pass
    return default_state()


def write_migration_state(state: dict[str, Any], state_dir: Path | None = None) -> None:
    path = migration_state_path(state_dir)
    path.parent.mkdir(parents=True, exist_ok=True)
    state = {**state, "updated_at": datetime.now(timezone.utc).isoformat()}
    tmp = path.with_suffix(f".tmp-{os.getpid()}")
    tmp.write_text(json.dumps(state, indent=2) + "\n")
    tmp.rename(path)


def mark_phase(
    phase: str | int,
    status: str,
    notes: str | None = None,
    state_dir: Path | None = None,
) -> dict[str, Any]:
    state = read_migration_state(state_dir)
    key = str(phase)
    record = dict(state.get("phases", {}).get(key, {"status": "pending"}))
    now = datetime.now(timezone.utc).isoformat()
    record["status"] = status
    if notes:
        record["notes"] = notes
    if status == "in_progress" and "started_at" not in record:
        record["started_at"] = now
    if status in ("complete", "rolled_back"):
        record["completed_at"] = now
    state.setdefault("phases", {})[key] = record
    if status == "complete":
        try:
            n = int(phase)
            if n >= state.get("current_phase", 1):
                state["current_phase"] = min(n + 1, 4)
        except ValueError:
            pass
    write_migration_state(state, state_dir)
    return state
